astropy/nddata/nddata.py

```diff
@@ -37,20 +30,15 @@ class NDData(NDDataBase):
     Parameters
     -----------
     data : `~numpy.ndarray`, `~numpy.ndarray`-like, or `NDData`
-        The actual data contained in this `NDData` object. If possible, data
-        will not be copied`data`, so you should make copy the ``data`` before
-        passing it in if that's the desired behavior.
+        The actual data contained in this `NDData` object.
 
     uncertainty : any type, optional
-        Uncertainty on the data. The uncertainty *must* have a string attribute
-        named ``uncertainty_type``, but there is otherwise no restriction.
+        Uncertainty on the data. The uncertainty *should* have a string
```

Comment:
I might be on your side but that's mostly equivalent on enforcing the uncertainty being a `NDUncertainty`. Since having an attribute requires a customized class and having uncertainty propagation requires that this class has an interface that matches the workflow of `ArithmeticMixin`. Since I'm not nearly finished with the arithmetic and propagation (propagation should work, but there are some unresolved issues regarding unit comparison I need to investigate) this requires the end-user or package that uses the `NDData` to emulate `NDUncertainty` or use it. 

But in some cases one might just consider saving some uncertainty sources there (maybe a dict) or the class just assumes standard-deviation or variance and works with that. I think issueing a warning when there is no such attribute should be enough. That forces someone who works with `NDData` to either endure the `log.Info`s or use `NDUncertainty` (or something like it). Currently if one wants to use some simpler form like just a  numpy array or dict one needs to subclass NDData and overwrite the uncertainty setter. With the `should` it is inconvieniant to use something else but one does not need any programming skills to use it.

I can revert the behaviour back to `must` but I really tried not to enforce anything but to suggest it instead.